superset-frontend/src/components/TableSelector/index.tsx

```diff
@@ -129,179 +175,173 @@ const TableSelector: FunctionComponent<TableSelectorProps> = ({
+          };
         });
-    }
-    setTableLoading(false);
-    setTableOptions([]);
-    return Promise.resolve();
-  }
+      }
+      return { data: [], totalCount: 0 };
+    },
+    // We are using the refresh state to re-trigger the query
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [currentDbId, currentSchema, onTablesLoad, schema, refresh],
+  );
 
   useEffect(() => {
-    if (dbId && schema) {
-      fetchTables();
+    async function fetchTable() {
+      if (schema && tableName) {
+        const table = await loadTable(dbId, schema, tableName);
+        if (table) {
+          setCurrentTable({
+            label: <TableOption table={table} />,
+            text: table.label,
+            value: table.value,
+          });
+        }
+      }
     }
-  }, [dbId, schema]);
+    fetchTable();
+  }, [dbId, loadTable, schema, tableName]);
 
   function onSelectionChange({
     dbId,
     schema,
-    tableName,
+    table,
   }: {
     dbId: number;
     schema?: string;
-    tableName?: string;
+    table?: TableOption;
   }) {
-    setCurrentTableName(tableName);
+    setCurrentTable(table);
+    setCurrentDbId(dbId);
     setCurrentSchema(schema);
     if (onUpdate) {
-      onUpdate({ dbId, schema, tableName });
+      onUpdate({ dbId, schema, tableName: table?.value });
     }
   }
 
-  function getTableNamesBySubStr(substr = 'undefined') {
-    if (!dbId || !substr) {
-      const options: any[] = [];
-      return Promise.resolve({ options });
-    }
-    const encodedSchema = encodeURIComponent(schema || '');
-    const encodedSubstr = encodeURIComponent(substr);
-    return SupersetClient.get({
-      endpoint: encodeURI(
-        `/superset/tables/${dbId}/${encodedSchema}/${encodedSubstr}`,
-      ),
-    }).then(({ json }) => {
-      const options = json.options.map((o: any) => ({
-        value: o.value,
-        schema: o.schema,
-        label: o.label,
-        title: o.title,
-        type: o.type,
-      }));
-      return { options };
-    });
-  }
-
-  function changeTable(tableOpt: any) {
-    if (!tableOpt) {
-      setCurrentTableName('');
+  function changeTable(table: TableOption) {
+    if (!table) {
+      setCurrentTable(undefined);
       return;
     }
-    const schemaName = tableOpt.schema;
-    const tableOptTableName = tableOpt.value;
-    if (tableNameSticky) {
+    const tableOptTableName = table.value;
+    if (currentDbId && tableNameSticky) {
       onSelectionChange({
-        dbId,
-        schema: schemaName,
-        tableName: tableOptTableName,
+        dbId: currentDbId,
+        schema: currentSchema,
+        table,
       });
     }
-    if (onTableChange) {
-      onTableChange(tableOptTableName, schemaName);
+    if (onTableChange && currentSchema) {
+      onTableChange(tableOptTableName, currentSchema);
     }
   }
 
-  function changeSchema(schemaOpt: any, force = false) {
-    const value = schemaOpt ? schemaOpt.value : null;
+  function onRefresh() {
```

Comment:
I think we need to add a force param here that the force refresh calls. Something like
```js
  function onRefresh(forceRefresh: boolean) {
```